src/passes/Inlining.cpp

```diff
@@ -143,22 +159,17 @@ struct FunctionInfo {
     if (options.shrinkLevel > 0 || options.optimizeLevel < 3) {
       return false;
     }
-    if (hasCalls) {
-      // This has calls. If it is just a trivial call itself then inline, as we
-      // will save a call that way - basically we skip a trampoline in the
-      // middle - but if it is something more complex, leave it alone, as we may
-      // not help much (and with recursion we may end up with a wasteful
-      // increase in code size).
-      //
-      // Note that inlining trivial calls may increase code size, e.g. if they
-      // use a parameter more than once (forcing us after inlining to save that
-      // value to a local, etc.), but here we are optimizing for speed and not
-      // size, so we risk it.
-      return isTrivialCall;
-    }
-    // This doesn't have calls. Inline if loops do not prevent us (normally, a
+    // The function just calls another function, but it's using locals in
+    // different order than the argument order, and/or using some locals more
+    // than once. In this case we inline if we're not optimizing for code size,
+    // as inlining it to more than one call site may increase code size by
+    // introducing locals.
+    if (trivialCall == TrivialCall::MayNotShrink) {
+      return true;
+    }
+    // Inline if it doesn't have calls and loops do not prevent us (normally, a
```

Comment:
I don't understand this sentence (up to the opening `(`). Is it

> Inline if it doesn't have calls and loops. Those do not prevent us

or

> Inline if it doesn't have calls. Loops do not prevent us

?

Reply:
This is an existing comment, I just removed the first sentence because it was no longer true with the changes.

What it's trying to say is: inline if (1) the function doesn't have calls (trivial calls are already handled at this point) and (2) the function doesn't have loops, or we're allowing inlining loops.

I'll try to rephrase, but the code is also clear enough (I think) so maybe it's OK to just drop the comment?